subprojects/language-java/src/main/java/org/gradle/api/internal/tasks/compile/incremental/deps/ClassSetAnalysis.java

```diff
@@ -258,37 +236,27 @@ private DependentsSet getDependents(String className) {
         }
         ImmutableSet<String> annotationProcessingClassDeps = classDependenciesFromAnnotationProcessing.get(className);
         ImmutableSet<GeneratedResource> annotationProcessingResourceDeps = resourceDependenciesFromAnnotationProcessing.get(className);
-        if (annotationProcessingClassDeps.isEmpty() && annotationProcessingResourceDeps.isEmpty()) {
-            return dependents;
-        }
-        return DependentsSet.dependents(dependents.getPrivateDependentClasses(),
-            Sets.union(dependents.getAccessibleDependentClasses(), annotationProcessingClassDeps),
-            Sets.union(dependents.getDependentResources(), annotationProcessingResourceDeps)
-        );
-    }
-
-    private DependentsSet getConstantDependents(String className) {
-        Set<String> accessibleConstantDependents = new ObjectOpenHashSet<>(compilerApiData.getConstantDependentsForClass(className).getAccessibleDependentClasses());
-        Set<String> privateConstantDependents = new ObjectOpenHashSet<>(compilerApiData.getConstantDependentsForClass(className).getPrivateDependentClasses());
-        processTransitiveConstantDependentClasses(new ObjectOpenHashSet<>(Collections.singleton(className)), privateConstantDependents, accessibleConstantDependents);
-        if (accessibleConstantDependents.contains(MODULE_INFO)) {
-            return DependentsSet.dependencyToAll("module-info has changed");
+        DependentsSet constantDependents = compilerApiData.getConstantDependentsForClass(className);
+        if (constantDependents.isDependencyToAll()) {
+            return constantDependents;
         }
-        return DependentsSet.dependents(privateConstantDependents, accessibleConstantDependents, Collections.emptySet());
-    }
-
-    private void processTransitiveConstantDependentClasses(Set<Object> visited, Set<String> privateConstantDependents, Set<String> accessibleConstantDependents) {
-        Deque<String> remainingAccessibleDependentClasses = new ArrayDeque<>(accessibleConstantDependents);
-        while (!remainingAccessibleDependentClasses.isEmpty()) {
-            String accessibleDependentClass = remainingAccessibleDependentClasses.pop();
-            if (!visited.add(accessibleDependentClass)) {
-                continue;
-            }
-            accessibleConstantDependents.add(accessibleDependentClass);
-            DependentsSet constantDependents = compilerApiData.getConstantDependentsForClass(accessibleDependentClass);
-            privateConstantDependents.addAll(constantDependents.getPrivateDependentClasses());
-            remainingAccessibleDependentClasses.addAll(constantDependents.getAccessibleDependentClasses());
+        if (annotationProcessingClassDeps.isEmpty() && annotationProcessingResourceDeps.isEmpty() && constantDependents.isEmpty()) {
+            return dependents;
         }
+        ImmutableSet.Builder<String> privateDependents = ImmutableSet.builder();
```

Comment:
All the different sources of dependencies are now handled consistently in this method, rather than being spread out over different code paths.